File: java/src/test/java/com/filefs/FileSystemTest.java
package com.filefs;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import org.junit.jupiter.api.Test;

class FileSystemTest {
    @Test
    void mkfsMountAndGetcwd() throws IOException {
        try (ImageFixture fixture = new ImageFixture("lifecycle")) {
            assertTrue(fixture.fs.isMounted());
            assertEquals("/", fixture.fs.getcwd());
        }
    }

    @Test
    void mkdirAndChdir() throws IOException {
        try (ImageFixture fixture = new ImageFixture("mkdir")) {
            fixture.fs.mkdir("docs");
            fixture.fs.chdir("docs");
            assertEquals("/docs/", fixture.fs.getcwd());
        }
    }

    @Test
    void openWriteReadRoundtrip() throws IOException {
        try (ImageFixture fixture = new ImageFixture("roundtrip")) {
            fixture.fs.mkdir("docs");
            fixture.fs.chdir("docs");

            FileHandle out = fixture.fs.open("note.txt", "w");
            byte[] payload = "hello filefs".getBytes();
            assertEquals(payload.length, fixture.fs.write(out, payload, 0, payload.length));
            fixture.fs.close(out);

            FileHandle in = fixture.fs.open("note.txt", "r");
            assertTrue(fixture.fs.seek(in, 0L, SeekWhence.END));
            assertEquals(payload.length, fixture.fs.tell(in));
            fixture.fs.rewind(in);

            byte[] read = new byte[64];
            int n = fixture.fs.read(in, read, 0, read.length);
            fixture.fs.close(in);

            assertEquals("hello filefs", new String(read, 0, n));
        }
    }

    @Test
    void copyRenameAndRemove() throws IOException {
        try (ImageFixture fixture = new ImageFixture("copy")) {
            FileHandle out = fixture.fs.open("orig.txt", "w");
            byte[] payload = "copy me".getBytes();
            assertEquals(payload.length, fixture.fs.write(out, payload, 0, payload.length));
            fixture.fs.close(out);

            fixture.fs.copyFile("orig.txt", "copy.txt");
            fixture.fs.rename("copy.txt", "renamed.txt");
            assertTrue(fixture.fs.fileExists("renamed.txt"));
            fixture.fs.removeFile("renamed.txt");
            assertFalse(fixture.fs.fileExists("renamed.txt"));
        }
    }

    @Test
    void openDirListsDocs() throws IOException {
        try (ImageFixture fixture = new ImageFixture("readdir")) {
            fixture.fs.mkdir("docs");
            DirectoryHandle dir = fixture.fs.openDir("/");
            assertNotNull(dir);
            assertEquals("/", dir.absolutePath());

            boolean foundDocs = false;
            while (true) {
                DirEntry entry = fixture.fs.readDir(dir);
                if (entry == null) {
                    break;
                }
                if ("docs".equals(entry.name()) && entry.type() == FileType.DIR) {
                    foundDocs = true;
                }
            }
            fixture.fs.closeDir(dir);
            assertTrue(foundDocs);
        }
    }

    @Test
    void beginCommitCreatesFile() throws IOException {
        try (ImageFixture fixture = new ImageFixture("txn")) {
            assertTrue(fixture.fs.begin());
            FileHandle out = fixture.fs.open("txn.txt", "w");
            byte[] payload = {'x'};
            assertEquals(1, fixture.fs.write(out, payload, 0, 1));
            fixture.fs.close(out);
            assertTrue(fixture.fs.commit());
            assertTrue(fixture.fs.fileExists("txn.txt"));
        }
    }

    private static final class ImageFixture implements AutoCloseable {
        final Path imagePath;
        final FileSystem fs = new FileSystem();

        ImageFixture(String name) throws IOException {
            imagePath = Files.createTempFile("filefs-java-" + name + "-", ".ffs");
            Files.deleteIfExists(imagePath);
            FileSystem.mkfs(imagePath);
            fs.mount(imagePath);
        }

        @Override
        public void close() throws IOException {
            fs.umount();
            Files.deleteIfExists(imagePath);
            Files.deleteIfExists(Path.of(imagePath.toString() + "-j"));
        }
    }
}

File: kotlin/src/test/kotlin/filefs/FileSystemTest.kt
package filefs

import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path

private fun assertTrue(condition: Boolean, message: String) {
    if (!condition) {
        throw AssertionError(message)
    }
}

private fun assertFalse(condition: Boolean, message: String) {
    assertTrue(!condition, message)
}

private fun assertEquals(expected: Any?, actual: Any?, message: String) {
    if (expected != actual) {
        throw AssertionError("$message expected=<$expected> actual=<$actual>")
    }
}

private fun assertNotNull(value: Any?, message: String) {
    if (value == null) {
        throw AssertionError(message)
    }
}

private class ImageFixture(name: String) : AutoCloseable {
    val imagePath: Path = Files.createTempFile("filefs-kotlin-$name-", ".ffs").also {
        Files.deleteIfExists(it)
    }
    val fs = FileSystem()

    init {
        FileSystem.mkfs(imagePath)
        fs.mount(imagePath)
    }

    override fun close() {
        fs.umount()
        Files.deleteIfExists(imagePath)
        Files.deleteIfExists(Path.of(imagePath.toString() + "-j"))
    }
}

private fun mkfsMountAndGetcwd() {
    ImageFixture("lifecycle").use { fixture ->
        assertTrue(fixture.fs.isMounted, "filesystem should be mounted")
        assertEquals("/", fixture.fs.getcwd(), "cwd should be root")
    }
}

private fun mkdirAndChdir() {
    ImageFixture("mkdir").use { fixture ->
        fixture.fs.mkdir("docs")
        fixture.fs.chdir("docs")
        assertEquals("/docs/", fixture.fs.getcwd(), "cwd should change to docs")
    }
}

private fun openWriteReadRoundtrip() {
    ImageFixture("roundtrip").use { fixture ->
        fixture.fs.mkdir("docs")
        fixture.fs.chdir("docs")

        val out = fixture.fs.open("note.txt", "w")
        val payload = "hello filefs".toByteArray(StandardCharsets.UTF_8)
        assertEquals(payload.size, fixture.fs.write(out, payload), "write should persist payload")
        fixture.fs.close(out)

        val input = fixture.fs.open("note.txt", "r")
        assertTrue(fixture.fs.seek(input, 0L, SeekWhence.END), "seek to end should succeed")
        assertEquals(payload.size.toLong(), fixture.fs.tell(input), "tell should report payload length")
        fixture.fs.rewind(input)

        val buffer = ByteArray(64)
        val count = fixture.fs.read(input, buffer)
        fixture.fs.close(input)

        assertEquals("hello filefs", String(buffer, 0, count, StandardCharsets.UTF_8), "roundtrip should match")
    }
}

private fun copyRenameAndRemove() {
    ImageFixture("copy").use { fixture ->
        val out = fixture.fs.open("orig.txt", "w")
        val payload = "copy me".toByteArray(StandardCharsets.UTF_8)
        assertEquals(payload.size, fixture.fs.write(out, payload), "write should persist source file")
        fixture.fs.close(out)

        fixture.fs.copyFile("orig.txt", "copy.txt")
        fixture.fs.rename("copy.txt", "renamed.txt")
        assertTrue(fixture.fs.fileExists("renamed.txt"), "renamed file should exist")
        fixture.fs.removeFile("renamed.txt")
        assertFalse(fixture.fs.fileExists("renamed.txt"), "removed file should not exist")
    }
}

private fun openDirListsDocs() {
    ImageFixture("readdir").use { fixture ->
        fixture.fs.mkdir("docs")
        val dir = fixture.fs.openDir("/")
        assertNotNull(dir, "directory handle should exist")
        assertEquals("/", dir.absolutePath(), "absolute path should be root")

        var foundDocs = false
        while (true) {
            val entry = fixture.fs.readDir(dir) ?: break
            if (entry.name == "docs" && entry.type == FileType.DIR) {
                foundDocs = true
            }
        }
        fixture.fs.closeDir(dir)
        assertTrue(foundDocs, "docs directory should be listed")
    }
}

private fun beginCommitCreatesFile() {
    ImageFixture("txn").use { fixture ->
        assertTrue(fixture.fs.begin(), "begin should succeed")
        val out = fixture.fs.open("txn.txt", "w")
        val payload = byteArrayOf('x'.code.toByte())
        assertEquals(1, fixture.fs.write(out, payload), "transactional write should succeed")
        fixture.fs.close(out)
        assertTrue(fixture.fs.commit(), "commit should succeed")
        assertTrue(fixture.fs.fileExists("txn.txt"), "committed file should exist")
    }
}

fun main() {
    val tests = listOf(
        "mkfsMountAndGetcwd" to ::mkfsMountAndGetcwd,
        "mkdirAndChdir" to ::mkdirAndChdir,
        "openWriteReadRoundtrip" to ::openWriteReadRoundtrip,
        "copyRenameAndRemove" to ::copyRenameAndRemove,
        "openDirListsDocs" to ::openDirListsDocs,
        "beginCommitCreatesFile" to ::beginCommitCreatesFile,
    )

    for ((name, test) in tests) {
        test()
        println("PASS $name")
    }
    println("PASS ${tests.size} tests")
}
